### ISSUE TITLE
Disabling `FAIL_ON_INVALID_SUBTYPE` breaks polymorphic deserialization of Enums

### ISSUE THREAD

**Issue by holgerknoche** [FasterXML/jackson-databind#2775]
Apparently, disabling the FAIL_ON_INVALID_SUBTYPE feature has some (at least for me) unexpected side-effects on the deserialization of enums. I originally found this issue in jackson-databind 2.10.0, but it still exists in 2.11.0. 
I have an enum that implements an interface:

    @JsonTypeInfo(use = Id.NAME)
    public interface CustomEnum {}

    @JsonTypeName("Test")
    public enum TestEnum implements CustomEnum {
        VALUE;
    }

Then, this test case works fine:

    ObjectMapper mapper = new ObjectMapper();
    mapper.registerSubtypes(TestEnum.class);
		
    mapper.enable(DeserializationFeature.FAIL_ON_INVALID_SUBTYPE);
		
    CustomEnum testValue = TestEnum.VALUE;
    String json = mapper.writeValueAsString(testValue);
		
    CustomEnum deserializedValue = mapper.readValue(json, CustomEnum.class);
    assertEquals(testValue, deserializedValue);

However, when I disable the feature, the test case fails because `deserializedValue` is suddenly `null`. The reason seems to be that a `NullyfingDeserializer` is returned from `TypeDeserializerBase._findDefaultImplDeserializer(...)` instead of `null`. Therefore, the `AsPropertyDeserializer` no longer tries to read the value itself.

**Comment by singleton11** [comment]
```
public class Test {

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
    interface CustomEnum {}

    @JsonTypeName("Test")
    enum TestEnum implements CustomEnum {
        VALUE;
    }

    @org.junit.Test
    public void test() throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerSubtypes(TestEnum.class);

//        mapper.enable(DeserializationFeature.FAIL_ON_INVALID_SUBTYPE);

        CustomEnum testValue = TestEnum.VALUE;
        String json = mapper.writeValueAsString(testValue);

        CustomEnum deserializedValue = mapper.readValue(json, CustomEnum.class);
        assertEquals(testValue, deserializedValue);
    }
}

```

This test works fine in 2.11 branch, what I do wrong?

**Comment by cowtowncoder** [comment]
Interesting. I can reproduce this. Workaround that trips handling appears to be fix for #148 so can not just remove use of `NullifyingDeserializer`.

**Comment by cowtowncoder** [comment]
Fundamentally the problem is that `Enum`s are serialized as `String`s -- scalar value -- and not as JSON Objects, so it is not possible use inclusion of `As.PROPERTY`. Instead, serialization occurs using `As.WRAPPER_ARRAY`.
But on deserialization this can not be determined. I am not sure why it appears to work with different settings however.

A solution here really is to use different inclusion:

```java
    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.WRAPPER_ARRAY)
```

and things will work.

It is odd that without disabling `FAIL_ON_INVALID_SUBTYPE` things appear to work ok, so I'll see if there is a way to handle this case as expected. But in the mean time specifying more general inclusion type (WRAPPER_ARRAY is guaranteed to work all shapes, ditto for WRAPPER_OBJECT).

**Comment by cowtowncoder** [comment]
I was able to fix this, but since there is some risk in fix will only add in for 2.12.0 (probably in 2.12.0-rc2), not for earlier minor versions. I suspect risk simply because this affects edge cases and testing is probably not extensive for such cases.
